README.md

```diff
@@ -211,42 +212,163 @@ val records = sc.newAPIHadoopFile(
+
+        ```scala
+        sc.hadoopConfiguration.set("fs.s3a.access.key", "YOUR_KEY_ID")
+        sc.hadoopConfiguration.set("fs.s3a.secret.key", "YOUR_SECRET_ACCESS_KEY")
+        ```
+
+        Python users will have to use a slightly different method to modify the `hadoopConfiguration`,
+        since this field is not exposed in all versions of PySpark. Although the following command
+        relies on some Spark internals, it should work with all PySpark versions and is unlikely to
+        break or change in the future:
+
+        ```python
+        sc._jsc.hadoopConfiguration().set("fs.s3n.awsAccessKeyId", "YOUR_KEY_ID")
+        sc._jsc.hadoopConfiguration().set("fs.s3n.awsSecretAccessKey", "YOUR_SECRET_ACCESS_KEY")
+        ```
+
+    3. **Encode keys in `tempdir` URI**:
+     For example, the URI `s3n://ACCESSKEY:SECRETKEY@bucket/path/to/temp/dir` encodes the key pair
+      (`ACCESSKEY`, `SECRETKEY`).
+
+      Due to [Hadoop limitations](https://issues.apache.org/jira/browse/HADOOP-3733), this
+      approach will not work for secret keys which contain forward slash (`/`) characters.
+
+- **Redshift to S3**: Redshift also connects to S3 during `COPY` and `UNLOAD` queries. There are
+    three methods of authenticating this connection:
+
+    1. **Have Redshift assume an IAM role (most secure)**: You can grant Redshift permission to assume
+        an IAM role during `COPY` or `UNLOAD` operations and then configure this library to instruct
+        Redshift to use that role:
+
+        1. Create an IAM role granting appropriate S3 permissions to your bucket.
+        2. Follow the guide
+        [_Authorizing Amazon Redshift to Access Other AWS Services On Your Behalf_](http://docs.aws.amazon.com/redshift/latest/mgmt/authorizing-redshift-service.html)
+        to configure this role's trust policy in order to allow Redshift to assume this role.
+        3. Follow the steps in the
+        [_Authorizing COPY and UNLOAD Operations Using IAM Roles_](http://docs.aws.amazon.com/redshift/latest/mgmt/copy-unload-iam-role.html)
+        guide to associate that IAM role with your Redshift cluster.
+        4. Set this library's `aws_iam_role` option to the role's ARN.
+    2. **Forward Spark's S3 credentials to Redshift**: if the `forward_spark_s3_credentials` option is
+        set to `true` then this library will automatically discover the credentials that Spark is
+        using to connect to S3 and will forward those credentials to Redshift over JDBC. If Spark
+        is authenticating to S3 using an IAM instance role then a set of temporary STS credentials
+        will be passed to Redshift; otherwise, AWS keys will be passed. These credentials are
+        sent as part of the JDBC query, so therefore it is **strongly recommended** to enable SSL
+        encryption of the JDBC connection when using this authentication method.
+    3. **Use Security Token Service (STS) credentials**: You may configure the
```

Comment:
These are also passed from the driver to redshift right? If so, shouldn't we also be encouraging TLS?